pkg/blockbuilder/blockbuilder.go

```diff
@@ -336,33 +321,38 @@ func (b *BlockBuilder) consumePartitions(ctx context.Context, part int32, mark t
 			}
 		})
 
-		if fetches.Empty() {
-			done = true
+		if fetches.Empty() && lag <= 0 {
+			level.Warn(b.logger).Log("msg", "got empty fetches from broker", "part", part)
+			break
 		}
 
+		var procFetchesErr error
 		fetches.EachPartition(func(ftp kgo.FetchTopicPartition) {
-			level.Info(b.logger).Log("msg", "consumed", "part", ftp.Partition, "hi", ftp.HighWatermark, "lo", ftp.LogStartOffset, "batch_size", len(ftp.Records))
+			level.Info(b.logger).Log("msg", "consumed", "part", ftp.Partition, "hi", ftp.HighWatermark, "lo", ftp.LogStartOffset, "lag", lag, "batch_size", len(ftp.Records))
+
+			lag -= int64(len(ftp.Records))
 
 			for _, rec := range ftp.Records {
 				// When BB is first deployed or if it is lagging behind, then it might consuming data from too much
 				// in the past. In which case if we try to consume all at once, it can overwhelm the system.
 				// So we break this into multiple block building cycles by resetting the block builder at intervals.
 				// TODO(codesome): the logic for this below is broken. Fix it. How can we determine the block ends when we are catching up?
-				if builder != nil && rec.Timestamp.After(resetBlockBuilderAt) {
+				if resetBlockBuilderAt.IsZero() {
+					resetBlockBuilderAt = rec.Timestamp.Truncate(b.cfg.ConsumeInterval).Add(b.cfg.ConsumeInterval + b.cfg.ConsumeIntervalBuffer)
+
+					// TODO(codesome): verify. this can be wrong.
+					currEnd = resetBlockBuilderAt.UnixMilli()
+					lastEnd = currEnd - b.cfg.ConsumeInterval.Milliseconds()
+				} else if rec.Timestamp.After(resetBlockBuilderAt) {
 					if err := builder.compactAndClose(ctx, b.shipperDir()); err != nil {
-						level.Error(b.logger).Log("msg", "failed to compact and remove dbs", "part", part, "err", err)
-						// TODO(codesome): return err?
 						// TODO(codesome): add metric
+						procFetchesErr = fmt.Errorf("compact and clean dbs: %w", err)
+						return
 					}
-					builder = nil
-				}
-				if builder == nil {
+					// TODO(v): maybe TSDBBuilder.compactAndClose should be doing compactAndFlush instead
 					builder = newTSDBBuilder(b.logger, b.limits, part, b.cfg.BlocksStorageConfig)
-					resetBlockBuilderAt = rec.Timestamp.Truncate(consumptionItvl).Add(consumptionItvl + timeBuffer)
-
-					// TODO(codesome): verify. this can be wrong.
-					currEnd = rec.Timestamp.Truncate(consumptionItvl).UnixMilli()
-					lastEnd = currEnd - consumptionItvl.Milliseconds()
+					// Zeroing the resetBlockBuilderAt marker to re-populate it on the next iteration.
```

Comment:
Since this record is after the `resetBlockBuilderAt`, we should be updating the currEnd and lastEnd here to process this record properly.

Reply:
I've refactored this part, PHAL